dvc/remote/base.py

```diff
@@ -628,22 +653,25 @@ def cache_exists(self, checksums):
         Returns:
             A list with checksums that were found in the remote
         """
-        progress_callback = ProgressCallback(len(checksums))
-
-        def exists_with_progress(chunks):
-            return self.batch_exists(chunks, callback=progress_callback)
-
-        if self.no_traverse and hasattr(self, "batch_exists"):
-            with ThreadPoolExecutor(max_workers=self.JOBS) as executor:
-                path_infos = [self.checksum_to_path_info(x) for x in checksums]
-                chunks = to_chunks(path_infos, self.JOBS)
-                results = executor.map(exists_with_progress, chunks)
-                in_remote = itertools.chain.from_iterable(results)
-                ret = list(itertools.compress(checksums, in_remote))
-                progress_callback.finish("")
-                return ret
-
-        return list(set(checksums) & set(self.all()))
+        if not self.no_traverse:
+            return list(set(checksums) & set(self.all()))
+
+        with ThreadPoolExecutor(max_workers=self.JOBS) as executor:
```

Comment:
what about the case when number of workers is dynamic (like we are maxing out the number of channels)

Reply:
If you are talking about sftp trick that we do in #2131, we would need to open as many connections as we can and then create a thread pool with that number of workers.

Reply:
so, self.JOBS won't be a constant, right? it might be confusing

Reply:
Maybe actually JOBS will be an upper limit. So if we can create JOBS connections - fine, if not - then we'll get what we'll get. Such an approach makes sense to me.

Reply:
map will max out number of workers immediately. what happens if `cxt_pool.get` in one of the workers fails to return a connection? how will pool scale itself down and reassign the work to other workers?